Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/consent/one/DccTicketingConsentOneViewModel.kt

```diff
@@ -1,31 +1,81 @@
 import de.rki.coronawarnapp.dccticketing.core.transaction.DccTicketingTransactionContext
+import de.rki.coronawarnapp.dccticketing.ui.shared.DccTicketingSharedViewModel
 import de.rki.coronawarnapp.util.coroutine.DispatcherProvider
 import de.rki.coronawarnapp.util.ui.SingleLiveEvent
+import de.rki.coronawarnapp.util.ui.toResolvingString
 import de.rki.coronawarnapp.util.viewmodel.CWAViewModel
 import de.rki.coronawarnapp.util.viewmodel.CWAViewModelFactory
 import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.map
+import timber.log.Timber
 
 class DccTicketingConsentOneViewModel @AssistedInject constructor(
-    @Assisted private val dccTicketingTransactionContext: DccTicketingTransactionContext,
-    dispatcherProvider: DispatcherProvider
+    @Assisted private val dccTicketingSharedViewModel: DccTicketingSharedViewModel,
+    dispatcherProvider: DispatcherProvider,
+    private val dccTicketingConsentOneProcessor: DccTicketingConsentOneProcessor
 ) : CWAViewModel(dispatcherProvider = dispatcherProvider) {
 
-    val showCloseDialog = SingleLiveEvent<Unit>()
+    private val currentIsLoading = MutableStateFlow(false)
+    val isLoading: LiveData<Boolean> = currentIsLoading.asLiveData()
 
-    val uiState = MutableStateFlow(UiState(dccTicketingTransactionContext)).asLiveData()
+    private val currentEvent = SingleLiveEvent<DccTicketingConsentOneEvent>()
+    val events: LiveData<DccTicketingConsentOneEvent> = currentEvent
 
-    fun goBack() {
-        showCloseDialog.postValue(Unit)
+    private val currentUiState = dccTicketingSharedViewModel.transactionContext
+        .map { UiState(it) }
+    val uiState: LiveData<UiState> = currentUiState.asLiveData2()
+
+    fun onUserCancel() {
+        Timber.d("onUserCancel()")
+        postEvent(ShowCancelConfirmationDialog)
+    }
+
+    fun goBack() = postEvent(NavigateBack)
+
+    fun showPrivacyInformation() = postEvent(NavigateToPrivacyInformation)
+
+    fun onUserConsent(): Unit = launch {
+        Timber.d("onUserConsent()")
+        currentIsLoading.compareAndSet(expect = false, update = true)
+        val event = try {
+            dccTicketingSharedViewModel.apply {
+                val ctx = transactionContext.first()
+                dccTicketingConsentOneProcessor.updateTransactionContext(ctx = ctx)
+                    .also { updateTransactionContext(ctx = it) }
+            }
+            NavigateToCertificateSelection
+        } catch (e: Exception) {
+            Timber.e(e, "Error while processing user consent")
+            val lazyErrorMessage = when (e) {
+                is DccTicketingException -> {
+                    val serviceProvider = currentUiState.first().provider
+                    e.errorMessage(serviceProvider = serviceProvider)
+                }
+                else -> R.string.errors_generic_text_unknown_error_cause.toResolvingString()
```

Comment:
Are the null checks going to be an unknown error?

Reply:
yes. They should only appear when a dev has done something "wrong"

Reply:
Like when someone forgot to set a value in the step before